nomad/resource_acl_auth_method.go

```diff
@@ -0,0 +1,407 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package nomad
+
+import (
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/nomad/api"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+)
+
+func resourceACLAuthMethod() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceACLAuthMethodCreate,
+		Update: resourceACLAuthMethodUpdate,
+		Delete: resourceACLAuthMethodDelete,
+		Read:   resourceACLAuthMethodRead,
+		Exists: resourceACLAuthMethodExists,
+
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description: "The identifier of the ACL Auth Method.",
+				Required:    true,
+				Type:        schema.TypeString,
+			},
+			"type": {
+				Description: `ACL Auth Method SSO workflow type. Currently, the only supported type is "OIDC."`,
+				Required:    true,
+				Type:        schema.TypeString,
+				ValidateFunc: validation.StringInSlice([]string{
+					api.ACLAuthMethodTypeOIDC,
+				}, false),
+			},
+			"token_locality": {
+				Description: `Defines whether the ACL Auth Method creates a local or global token when performing SSO login. This field must be set to either "local" or "global".`,
+				Required:    true,
+				Type:        schema.TypeString,
+				ValidateFunc: validation.StringInSlice([]string{
+					api.ACLAuthMethodTokenLocalityGlobal,
+					api.ACLAuthMethodTokenLocalityLocal,
+				}, false),
+			},
+			"max_token_ttl": {
+				Description: "Defines the maximum life of a token created by this method.",
+				Required:    true,
+				Type:        schema.TypeString,
+			},
+			"default": {
+				Description: "Defines whether this ACL Auth Method is to be set as default.",
+				Optional:    true,
+				Type:        schema.TypeBool,
+			},
+			"config": {
+				Description: "",
+				Required:    true,
+				Type:        schema.TypeList,
+				MaxItems:    1,
+				MinItems:    1,
+				Elem:        resourceACLAuthMethodConfig(),
+			},
+		},
+	}
+}
+
+func resourceACLAuthMethodConfig() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"oidc_discovery_url": {
+				Description: "",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"oidc_client_id": {
+				Description: "",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"oidc_client_secret": {
+				Description: "",
+				Type:        schema.TypeString,
+				Required:    true,
+				Sensitive:   true,
+			},
+			"oidc_scopes": {
+				Description: "",
+				Type:        schema.TypeList,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Optional:    true,
+			},
```

Comment:
Would it make sense to have an `oidc` block instead of a flat config? So a config would look like this:

```hcl
resource "nomad_acl_auth_method" "okta" {
  name = "okta"
  
  oidc {
    discovery_url = "..."
    client_id     = "..."
  }
}
```

We may be able to even remove the `type` 🤔

Reply:
It's not a bad shout; I don't know how we feel about changing the layout compared to the API. I'll play around with changing it, when I have a moment.

Reply:
I am inclined to keep this the same shape as the API and make things easier to translate between the two.